src/RoboSaverVirtualModuleFactory.sol

```diff
@@ -0,0 +1,124 @@
+                                  EXTERNAL METHODS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Creates a virtual module for a card
+    /// @param _delayModule The address of the delay module
+    /// @param _roleModule The address of the role module
+    /// @param _buffer The buffer for the virtual module (configurable)
+    /// @param _slippage The slippage for the virtual module (configurable)
+    function createVirtualModule(address _delayModule, address _roleModule, uint256 _buffer, uint16 _slippage)
+        external
+    {
+        // @todo sanity checks on the inputs!
+
+        // uses `CARD` address has to helps pre-determining the address of the virtual module given the salt
+        address virtualModule = address(
+            new RoboSaverVirtualModule{salt: keccak256(abi.encodePacked(msg.sender))}(
+                address(this), _delayModule, _roleModule, _buffer, _slippage
+            )
+        );
+
+        uint256 upkeepId = _registerRoboSaverVirtualModule(virtualModule);
+
+        // extracts forwarder address & sets keeper
+        address keeper = CL_REGISTRY.getForwarder(upkeepId);
+        RoboSaverVirtualModule(virtualModule).setKeeper(keeper);
+
+        emit RoboSaverVirtualModuleCreated(virtualModule, msg.sender, upkeepId, block.timestamp);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                   INTERNAL METHODS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Registers the virtual module in the Chainlink Keeper Registry
+    /// @param _virtualModule The address of the virtual module to be registered
+    function _registerRoboSaverVirtualModule(address _virtualModule) internal returns (uint256 upkeepId_) {
+        IKeeperRegistrar.RegistrationParams memory registrationParams = IKeeperRegistrar.RegistrationParams({
+            name: string.concat(RoboSaverVirtualModule(_virtualModule).name(), "-", _addressToString(msg.sender)),
+            encryptedEmail: "",
+            upkeepContract: _virtualModule,
+            gasLimit: 2_000_000, // @todo optimise from the gas logs the right value. perhaps we are overshooting
+            adminAddress: address(this), // @note the factory is the admin
+            triggerType: 0,
+            checkData: "",
+            triggerConfig: "",
+            offchainConfig: "",
+            amount: 200e18 // @note dummy value for now
+        });
+
+        upkeepId_ = CL_REGISTRAR.registerUpkeep(registrationParams);
+        if (upkeepId_ == 0) revert UpkeepZero();
+
+        virtualModules[msg.sender] = VirtualModuleDetails({virtualModuleAddress: _virtualModule, upkeepId: upkeepId_});
+    }
+
+    /// @notice Converts an address to a string
+    function _addressToString(address _addr) internal pure returns (string memory) {
```

Comment:
this function was already somewhere no? or am i confused now

Reply:
yes, it was defined in the base fixture, but we need it here as well. i think we can do a clean up a later point of methods we use multiple times when we reorg the constants, event etc

Reply:
ah right